src/main/resources/avro/references.avdl

```diff
@@ -1,22 +1,74 @@
 @namespace("org.ga4gh.models")
 
+/**
+Defines types used by the GA4GH References API.
+
+References API servers may implement either or both of two modes: "classic", in
+which all `Reference`s annotate simple linear strings, and "graph", in which
+`Reference`s annotate sequences in sequence graphs. The "graph" mode is to be
+prefered in new client and server implementations.
+*/
 protocol References {
 
+import idl "common.avdl";
+
 /**
 A `Reference` is a canonical assembled contig, intended to act as a
 reference coordinate space for other genomic annotations. A single
 `Reference` might represent the human chromosome 1, for instance.
+
+An API server that supports the "graph" mode must always set the `segment`
+field. An API server that supports the "classic" mode must always set the `id`
+field. New client and server implementations should prefer the "graph" mode.
+
 */
 record Reference {
-  /** The reference ID. Unique within the repository. */
-  string id;
+
+  /**
+  The `Segment` describing this `Reference`'s sequence in the `ReferenceSet`'s
+  sequence graph.
+
+  The `Segment` will generally cover the entirety of the sequence, but may be
+  shorter if only a subrange of a sequence is to be included in the
+  `ReferenceSet`.
+
+  The sequence ID of the `Segment` is equal to ID of the `Reference`. The
+  actual sequence bases for a `Reference` are available through the
+  `getSequence()` API call on this `Reference`'s sequence.
+
+  If the API server supports the "graph" mode, this field must not be null.
+  */
+  union { null, Segment } segment = null;
+
+  /**
+  The reference ID. Unique within the repository. A sequence with this ID
+  exists, containing the reference's bases. If null, `segment` must be set, and
+  the sequence ID field from `segment` is used.
+
+  If the API server supports the "classic" mode, this field must not be null.
+  */
+  union { null, string } id = null;
 
   /** The length of this reference's sequence. */
   long length;
 
   /**
-  MD5 of the upper-case sequence excluding all whitespace characters
-  (this is equivalent to SQ:M5 in SAM).
+  The MD5 checksum uniquely representing this `Reference` and its position in
+  the `ReferenceSet`'s sequence graph, as a lower-case hexadecimal string.
+
+  If `segment.startJoin` and `segment.endJoin` are both null, or `segment`
+  itself is null, this is the MD5 of the upper-case sequence excluding all
+  whitespace characters (this is equivalent to SQ:M5 in SAM).
+
+  Otherwise, this is the MD5 of the above MD5 checksum, `segment.startJoin`'s
```

Comment:
It's not referring to the `md5checksum` of the associated joined segment; it's referring to the MD5 of the joined segment's `md5checksum` concatenated with the join position as a decimal string and a "+" or "-" character denoting the strand of the join.

I've rewritten this section entirely to specify exactly what you are supposed to stir into your hash. Please let me know if it's still unclear.